### PR TITLE
Fix app_data accessiblity in middlewares

### LINKED ISSUE #1790: Cannot access app_data in middleware of Scope
## Simple Use Case
```rust
...
web::scope("/api")
    .app_data(web::Data::new(1usize))
    .wrap_fn(|req, srv| {
        let i = req.app_data::<usize>();
        println!("i {} ", i.is_some()); // outputs "i false"
        let fut = srv.call(req);

        async move {
            let res = fut.await?;
            Ok(res)
        }
    })
...
```
## Expected Behavior
Ability to access the request's `app_data` within the middleware (`wrap` / `wrap_fn`) of a `Scope`.

## Current Behavior
Looking at the example code above, you would expect that `app_data` from the request can be retrieved, but it does not work.
If you move the setting of `app_data` outside the scope (ex. on `App::new().app_data(data)`) the data is accessible from within the middleware.

## Possible Solution
Spoke with @robjtede on the Gitter channel and he advised me to submit this issue. I am currently looking into debugging it myself, but I am new to the code.

One thought that I have is that in the use case above you can't get the data because `ScopeService` doesn't invoke `call()` until after trying to get the data, since on `ScopeService::call()` it stuffs its data into the request.

* Rust Version (I.e, output of `rustc -V`): rustc 1.50.0-nightly (825637983 2020-11-18)
* Actix Web Version: actix-web = "3"

### PR DESCRIPTION
**Pull request by aliemjay**
## PR Type

Bug fix

## PR Checklist

- [x] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt.
- [x] (Team) Label with affected crates and semver status.

## Overview

An alternative implementation for #2261. This delegates the responsility for setting `app_data` to upstream routing services (`App` and `Scope`). Since `Resource` is not a router, it doesn't have to set `app_data` itself. 

Closes #1790 
Closes #2261

### CODE REVIEW

**Comment by aliemjay** [comment]
Giving it more thought, I think #2261 still have some merit as less invasive change.

File: src/scope.rs [hunk 1]
```diff
@@ -537,11 +542,10 @@ impl Service<ServiceRequest> for ScopeService {
...
-        }
-
-        if let Some((srv, _info)) = res {
+        if let Some(((srv, app_data), _info)) = res {
+            if let Some(ref app_data) = app_data {
+                req.add_data_container(app_data.clone());
+            }
             srv.call(req)
         } else {
             self.default.call(req)
```
Comment (robjtede) [inline src/scope.rs:553]: there's a regression here in that default services are losing access to it's layer's data

Comment (aliemjay) [reply src/scope.rs:553]: Its layer's data are already set by  the parent layer before calling Service::call

Comment (aliemjay) [reply src/scope.rs:553]: At leat this is what I think. I'll write a test case for this once I get to my dev pc.

Comment (aliemjay) [reply src/scope.rs:553]: There are already some unit tests for such case: `resource::tests::test_data_default_service` and `scope::tests::test_override_data_default_service`.

So I can confirm that this behaves as I expect: any layer's app_data are registered to and set by the parent layer's routing service.  In other words, `req.add_data_container(app_data.clone());`, in the quoted diff, sets app_data for lower layers, not the current one.

**Comment by aliemjay** [comment]
Closing in favor of #2288

**Pull request closed without merging**